### PR TITLE
Add Scalafix rewrites for 1.0.0

### PR DESCRIPTION
**Pull request by gabro**
This is a WIP for #1791.

It's very preliminary, but feel free to share your thoughts, try this out and add more edge cases to the tests!

### CODE REVIEW

**Comment by kailuowang** [comment]
This is awesome. I suggest that we make it available to more user sooner. 
If we can add a quick step by step instruction on how to use this, we can merge it now and link it from the migration guide. It doesn't have to include migrations for all the breaking changes. We can add those incrementally - that would also allow more people to contribute. We just need to communicate that this is an incomplete auto-migration. 
What do you think?

**Comment by gabro** [comment]
I pushed another easy one (`RenameFreeSuspend`) and refactored some code a bit.

@kailuowang I agree that merging soon would be a good idea. About the strategy to make this available to the users I would like @olafurpg's opinion, since I think this is the first real-world scenario of making a rewrite available with a library and things are still a bit sketchy.

**Comment by olafurpg** [comment]
Great work @gabro!

The simplest way to share a rewrite at the moment is to put it in a single source file under `scalafix/rewrites/src/main/scala/Cats_v1.scala` and users can run it from the sbt shell with `> scalafix github:typelevel/cats/v1` with the sbt-scalafix plugin. It's possible to download the file locally and run `> scalafix file:path/to/Cats_v1.scala` if people prefer that.

Alternatively, it's possible to publish the rewrites to Maven. As long as the artifacts are on the same classpath as scalafix then users can run the rewrites with `> scalafix scala:fix.Cats_v1` (assuming rewrite fqn is `fix.Cats_v1`).

The sbt plugin is a thin shim over the scalafix cli so it should be possible to integrate with other build tools than sbt. Pants has a scalafix integration, but I'm not aware of plugins for gradle or maven.

I hope to release scalafix 0.5.0 later this week, I addressing a few issues in the sbt plugin first.

**Comment by kailuowang** [comment]
thanks @gabro! I agree `sbt-scalafix` running using github src is the easiest.

**Comment by gabro** [comment]
I just had a chat with @olafurpg about the best strategy for this, and we came up with the following solution (using scalafix 0.5):

- we'll keep all the rewrites separated, so that people can run them in isolation in case anything goes wrong
- I'll generate a single file called `Cats_v1_0_0.scala` (or something like that) that's just the concatenation of all the single rewrites in a single file. This is currently the most convenient way of running all the rewrites with scalafix, although it involves duplicating the source (single rewrites + generated file)

This will allow users to run all the rewrites using the scalafix sbt plugin: `sbt scalafix github:typelevel/cats/v1_0_0` 🎉 

We also drafted a solution for future versions of scalafix for making the process much simpler for both the rewrite authors and the users. @olafurpg will provide more details soon :)

**Comment by gabro** [comment]
My bad, I got some parts of that mixed up. There won't be any duplication, just a single `Cats_v1_0_0.scala` file containing all the rewrites. Coming soon :)

**Comment by olafurpg** [comment]
Any recommendations for OSS repos where we can try out the migration?

**Comment by kailuowang** [comment]
https://github.com/stew/dogs might be a good candidate.

**Comment by gabro** [comment]
@kailuowang I've covered a few corner case of the available rewrites and added some instructions on how to run the fixes. Take a look ;)

**Comment by gabro** [comment]
Also, once this gets merged I would suggest to use the original issue as an aggregation point for bug reports and general feedback. I expect *a lot* of corner cases will be found.

**Comment by kailuowang** [comment]
Looking great to me, thanks so much! :+1:

**Comment by kailuowang** [comment]
Would love to get this merge and start to benefit users and iterate.

**Comment by milessabin** [comment]
LGTM. I'd like to see this merged ASAP :-)

File: scalafix/project/plugins.sbt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+addSbtPlugin("ch.epfl.scala" % "sbt-scalafix" % "0.5.0-M1")
```
Comment (pepegar) [inline scalafix/project/plugins.sbt]: This line is repeated :)

File: scalafix/project/plugins.sbt [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+addSbtPlugin("ch.epfl.scala" % "sbt-scalafix" % "0.5.0-M1")
```
Comment (peterneyens) [inline scalafix/project/plugins.sbt]: Once is good enough, right?

Comment (gabro) [inline scalafix/project/plugins.sbt]: most definitely! Good catch, thanks :)

File: scalafix/project/plugins.sbt (hunk 1 above)
Comment (gabro) [inline scalafix/project/plugins.sbt]: good catch, thank you!

**Comment by gabro** [comment]
Updated, I've removed the duplicated line in plugins and bumped scalafix to the latest 0.5.0-M2 version.
